File: LeetCode/Easy/0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        char[] ch = s.toCharArray();
        Deque<Character> stack = new ArrayDeque<>();
        
        for (char c : ch) {
            if (c == ')') {
                if (!stack.isEmpty() && stack.peek() == '(') {
                    stack.pop();
                } else {
                    return false;
                }
            } else if (c == ']') {
                if (!stack.isEmpty() && stack.peek() == '[') {
                    stack.pop();
                } else {
                    return false;
                }
            } else if (c == '}') {
                if (!stack.isEmpty() && stack.peek() == '{') {
                    stack.pop();
                } else {
                    return false;
                }
            } else {
                stack.push(c);
            }
        }
        
        return stack.isEmpty() ? true : false;
    }
}

File: LeetCode/Easy/0020-valid-parentheses/0020-valid-parentheses.kt
class Solution {
    fun isValid(s: String): Boolean {
        val stack = ArrayDeque<Char>()

        for (c in s) {
            when (c) {
                '(', '[', '{' -> stack.addFirst(c)
                ')' -> if (stack.removeFirstOrNull() != '(') return false
                ']' -> if (stack.removeFirstOrNull() != '[') return false
                '}' -> if (stack.removeFirstOrNull() != '{') return false
            }
        }

        return stack.isEmpty()
    }
}